java/ql/test/library-tests/frameworks/spring/util/Test.java

```diff
@@ -0,0 +1,1445 @@
+			File in = (File)source();
+			FileSystemUtils.copyRecursively(in, out);
+			sink(out); // $hasTaintFlow
+		}
+		{
+			// "org.springframework.util;LinkedMultiValueMap;false;LinkedMultiValueMap;(java.util.Map);;Element of MapValue of Argument[0];Element of MapValue of Argument[-1];value"
+			LinkedMultiValueMap out = null;
+			Map in = Map.of(null, List.of(source()));
+			out = new LinkedMultiValueMap(in);
+			sink(getElement((List)getMapValue(out))); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;LinkedMultiValueMap;false;LinkedMultiValueMap;(java.util.Map);;MapKey of Argument[0];MapKey of Argument[-1];value"
+			LinkedMultiValueMap out = null;
+			Map in = Map.of(source(), null);
+			out = new LinkedMultiValueMap(in);
+			sink(getMapKey(out)); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;LinkedMultiValueMap;false;deepCopy;;;MapKey of Argument[-1];MapValue of ReturnValue;value"
+			LinkedMultiValueMap out = null;
+			LinkedMultiValueMap in = (LinkedMultiValueMap)Map.of(source(), null);
+			out = in.deepCopy();
+			sink(getMapValue(out)); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;LinkedMultiValueMap;false;deepCopy;;;MapValue of Argument[-1];MapValue of ReturnValue;value"
+			LinkedMultiValueMap out = null;
+			LinkedMultiValueMap in = (LinkedMultiValueMap)Map.of(null, source());
+			out = in.deepCopy();
+			sink(getMapValue(out)); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;MultiValueMap;true;add;;;Argument[0];MapKey of Argument[-1];value"
+			MultiValueMapAdapter out = null;
+			Object in = (Object)source();
+			out.add(in, null);
+			sink(getMapKey(out)); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;MultiValueMap;true;add;;;Argument[0];MapKey of Argument[-1];value"
+			MultiValueMap out = null;
+			Object in = (Object)source();
+			out.add(in, null);
+			sink(getMapKey(out)); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;MultiValueMap;true;add;;;Argument[0];MapKey of Argument[-1];value"
+			MultiValueMap out = null;
+			Object in = (Object)source();
+			out.add(in, (Object)null);
+			sink(getMapKey(out)); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;MultiValueMap;true;add;;;Argument[1];Element of MapValue of Argument[-1];value"
+			MultiValueMapAdapter out = null;
+			Object in = (Object)source();
+			out.add(null, in);
+			sink(getElement((List)getMapValue(out))); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;MultiValueMap;true;add;;;Argument[1];Element of MapValue of Argument[-1];value"
+			MultiValueMap out = null;
+			Object in = (Object)source();
+			out.add(null, in);
+			sink(getElement((List)getMapValue(out))); // $hasValueFlow
+		}
+		{
+			// "org.springframework.util;MultiValueMap;true;add;;;Argument[1];Element of MapValue of Argument[-1];value"
+			MultiValueMap out = null;
+			Object in = (Object)source();
+			out.add((Object)null, in);
+			sink(getElement((List)getMapValue(out))); // $hasValueFlow
```

Comment:
All of these casts to `List` can be removed with the above sketched improvement to `getMapValue`.